psydac/api/tests/test_api_feec_1d.py

```diff
@@ -221,21 +211,34 @@ class CollelaMapping1D(Mapping):
     # Scalar diagnostics setup
     #--------------------------------------------------------------------------
 
-    # Energy of exact solution
-    def exact_energies(t, E_ex, B_ex):
-        """ Compute electric & magnetic energies of exact solution.
-        """
-        We = 0.5 * quad(lambda x: E_ex(t, x)**2, xmin, xmax)[0]
-        Wb = 0.5 * quad(lambda x: B_ex(t, x)**2, xmin, xmax)[0]
-        return (We, Wb)
-
-    # Energy of numerical solution
-    def discrete_energies(e, b):
-        """ Compute electric & magnetic energies of numerical solution.
-        """
-        We = 0.5 * M0.dot(e).dot(e)
-        Wb = 0.5 * M1.dot(b).dot(b)
-        return (We, Wb)
+    class Diagnostics:
+
+        def __init__(self, E_ex, B_ex, M0, M1):
+            self._E_ex = E_ex
+            self._B_ex = B_ex
+            self._M0 = M0
+            self._M1 = M1
+            self._tmp0 = None
+            self._tmp1 = None
+
+        # Energy of exact solution
+        def exact_energies(self, t):
+            """ Compute electric & magnetic energies of exact solution.
+            """
+            We = 0.5 * quad(lambda x: self._E_ex(t, x)**2, xmin, xmax)[0]
+            Wb = 0.5 * quad(lambda x: self._B_ex(t, x)**2, xmin, xmax)[0]
+            return (We, Wb)
+
+        # Energy of numerical solution
+        def discrete_energies(self, e, b):
+            """ Compute electric & magnetic energies of numerical solution.
+            """
+            # Why does one need "self._tmp0 = "?
+            self._tmp0 = self._M0.dot(e, out=self._tmp0)
+            self._tmp1 = self._M1.dot(b, out=self._tmp1)
+            We = 0.5 * self._tmp0.dot(e)
+            Wb = 0.5 * self._tmp1.dot(b)
+            return (We, Wb)
```

Comment:
Q:
Why does one need "self._tmp0 = "?

A:
When an object of class `Diagnostics` is created, we have that `self._tmp0 = None`. Hence the first call to `discrete_energies` will allocate a vector by executing the line
```python
self._tmp0 = self._M0.dot(e, out=None)
```
where it is clearly fundamental to assign the result to `self._tmp0`. In the successive calls to `discrete_energies` the assignment is superfluous, but should not cause any noticeable overhead.

Reply:
Alternatively one could allocate the temporary arrays `self._tmp0` and `self._tmp1` in the `__init__` constructor.

Reply:
Thank you for clarifying!